File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    // head recursion
    void reverseListHelper(ListNode* curr, ListNode* prev, ListNode* &head) {
        if(curr == NULL) {
            // update head which we need to return 
            head = prev;
            return;
        }
        
        ListNode* next = curr->next;
        curr -> next = prev;
        
        reverseListHelper(next, curr, head);
    }
    
    ListNode* reverseList(ListNode* &head) {
        ListNode* curr = head;
        ListNode* prev = NULL;
        
        reverseListHelper(curr, prev, head);
        
        return head;
    }
};

// tc -> O(n)
// sc -> O(n) // for stack 

File: 0206-reverse-linked-list/0206-reverse-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    // check if linked list is empty of have one element
    if(head == nil || head.Next == nil) {
        return head
    } 
    
    cNode := head
    var pNode *ListNode // assignd nil rn
    
    for cNode != nil {
        temp := cNode.Next
        cNode.Next = pNode
        pNode = cNode
        cNode = temp
    }
    
    return pNode
}